firedrake/matrix_free/preconditioners.py

```diff
@@ -0,0 +1,311 @@
+
+        """
+        pass
+
+
+class AssembledPC(PCBase):
+    """A matrix-free PC that assembles the operator.
+
+    Internally this makes a PETSc PC object that can be controlled by
+    options using the extra options prefix ``assembled_``.
+    """
+    def initialize(self, pc):
+        from firedrake.assemble import assemble
+
+        _, P = pc.getOperators()
+
+        context = P.getPythonContext()
+
+        # It only makes sense to preconditioner/invert a diagonal
+        # block in general.  That's all we're going to allow.
+        if not context.on_diag:
+            raise ValueError("Only makes sense to invert diagonal block")
+
+        self.P = assemble(context.a, bcs=context.row_bcs,
+                          form_compiler_parameters=context.fc_params, nest=False)
+        self.P.force_evaluation()
+
+        # Transfer nullspace over
+        Pmat = self.P.petscmat
+        Pmat.setNullSpace(P.getNullSpace())
+        Pmat.setTransposeNullSpace(P.getTransposeNullSpace())
+
+        prefix = pc.getOptionsPrefix()
+
+        # Internally, we just set up a PC object that the user can configure
+        # however from the PETSc command line.  Since PC allows the user to specify
+        # a KSP, we can do iterative by -assembled_pc_type ksp.
+        pc = PETSc.PC().create()
+        pc.setOptionsPrefix(prefix+"assembled_")
+        pc.setOperators(Pmat, Pmat)
+        pc.setUp()
+        pc.setFromOptions()
+        self.pc = pc
+
+    def update(self, pc):
+        from firedrake import assemble
+        P = self.P
+        P = assemble(P.a, tensor=P, bcs=P.bcs)
+        P.force_evaluation()
+
+    def apply(self, pc, x, y):
+        self.pc.apply(x, y)
+
+    def applyTranspose(self, pc, x, y):
+        self.pc.applyTranspose(x, y)
+
+    def view(self, pc, viewer=None):
+        super(AssembledPC, self).view(pc, viewer)
+        viewer.printfASCII("PC to apply inverse\n")
+        viewer.pushASCIITab()
+        self.pc.view(viewer)
+        viewer.popASCIITab()
+
+
+class MassInvPC(PCBase):
+
+    """A matrix free operator that inverts the mass matrix in the provided space.
+
+    Internally this creates a PETSc KSP object that can be controlled
+    by options using the extra options prefix ``Mp_``.
+    """
+    def initialize(self, pc):
+        from firedrake import TrialFunction, TestFunction, dx, assemble, inner
+        prefix = pc.getOptionsPrefix()
+
+        # we assume P has things stuffed inside of it
+        _, P = pc.getOperators()
+        context = P.getPythonContext()
+
+        test, trial = context.a.arguments()
+
+        if test.function_space() != trial.function_space():
+            raise ValueError("MassInvPC only makes sense if test and trial space are the same")
+
+        V = test.function_space()
+
+        u = TrialFunction(V)
+        v = TestFunction(V)
+        # Handle vector and tensor-valued spaces.
+        a = inner(u, v)*dx
```

Comment:
My stokes example had unit viscosity.  But I agree that this could be more
general and could come in via extra context.
But I thought GMRES was scale-invariant (i.e. scaling the
preconditioner/system doesn't affect convergence rate.)


On Mon, Aug 15, 2016 at 9:11 AM, Lawrence Mitchell <notifications@github.com
> wrote:

> In firedrake/matrix_free/preconditioners.py
> <https://github.com/firedrakeproject/firedrake/pull/853#discussion_r74766711>
> :
>
> > +
> > +        # we assume P has things stuffed inside of it
> > +        _, P = pc.getOperators()
> > +        context = P.getPythonContext()
> > +
> > +        test, trial = context.a.arguments()
> > +
> > +        if test.function_space() != trial.function_space():
> > +            raise ValueError("MassInvPC only makes sense if test and trial space are the same")
> > +
> > +        V = test.function_space()
> > +
> > +        u = TrialFunction(V)
> > +        v = TestFunction(V)
> > +        # Handle vector and tensor-valued spaces.
> > +        a = inner(u, v)*dx
>
> For use in stokes, one should use the viscosity weighted mass-matrix 1/mu
> u*v*dx for spectral equivalence for low-contrast viscosities. I guess
> this should come in through the context argument as for the navier-stokes
> reynolds number.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/firedrakeproject/firedrake/pull/853/files/ff5ec4e0eabb94576ac69c9c82932eaa50b2d14e#r74766711>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADl6L0eAiPfKnjnMwh-Ho2qBAQgHWmziks5qgHOSgaJpZM4JiV-g>
> .
>